src/compiler/checker.ts

```diff
@@ -8590,27 +8612,17 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                         visitNode(node.default, visitExistingNodeTreeSymbols, isTypeNode),
                     );
                 }
-                if (isTypeReferenceNode(node)) {
-                    if (canReuseTypeNode(context, node)) {
-                        const { introducesError, node: newName } = trackExistingEntityName(node.typeName, context);
-                        const typeArguments = visitNodes(node.typeArguments, visitExistingNodeTreeSymbols, isTypeNode);
-
-                        if (!introducesError) {
-                            const updated = factory.updateTypeReferenceNode(
-                                node,
-                                newName,
-                                typeArguments,
-                            );
-                            return setTextRange(context, updated, node);
-                        }
-                        else {
-                            const serializedName = serializeTypeName(context, node.typeName, /*isTypeOf*/ false, typeArguments);
-                            if (serializedName) {
-                                return setTextRange(context, serializedName, node.typeName);
-                            }
-                        }
+
+                if (isIndexedAccessTypeNode(node) && isTypeReferenceNode(node.objectType)) {
+                    const objectType = tryVisitTypeReference(node.objectType);
+                    if (!objectType) {
+                        return serializeExistingTypeNode(context, node);
                     }
-                    return serializeExistingTypeNode(context, node);
+                    return factory.updateIndexedAccessTypeNode(node, objectType, visitNode(node.indexType, visitExistingNodeTreeSymbols, isTypeNode)!);
+                }
+
+                if (isTypeReferenceNode(node)) {
+                    return tryVisitTypeReference(node) ?? serializeExistingTypeNode(context, node);

```

Comment:
Are the explicit fallbacks needed with the automatic fallback added in https://github.com/microsoft/TypeScript/pull/58720?

Reply:
Yeah. You are right. Better to lean on the new fallback mechanism. As it currently is we could end up trying to print the type two times (when we fail). Pushed the change